### PR TITLE
Change return type of QrCode query to better indicate errors

### PR DESCRIPTION
**Pull request by pylipp**
The unions may contain error types with detailed info. I hope the following table clarifies how to deal with the response of the query https://github.com/boxwise/boxtribute/blob/change-qrcode-query-interface/front/src/queries/queries.ts#L38-L66

| Union fragment | Only set if ... | Related current response | Current handling in FE |
| --- | --- | --- | --- |
| 1. QrCodeResult.InsufficientPermissionError | ...user misses the `qr:read` permission (e.g. a label-creation user) | error code `FORBIDDEN` | message "You don't have permission to access this box!" |
| 2. QrCodeResult.ResourceDoesNotExistError | ...the QR code for the requested hash does not exist in the database | error code `BAD_USER_INPUT` | message "No box found for this QR code!" |
| 3. QrCodeResult.QrCode | ...the errors 1.&2. don't occur | QR code data | success; redirect accordingly |
| 3a. QrCodeResult.BoxResult.InsufficientPermissionError | ...if 3. happens AND user misses the `stock:read` permission | like in 1. | like in 1. |
| 3b. QrCodeResult.BoxResult.UnauthorizedForBaseError | ...if 3. happens AND user has `stock:read` permission BUT NOT for the base that the requested box is in | like in 1. | like in 1.
| 3c. QrCodeResult.BoxResult.Box | ...if the errors 3a. and 3b. don't occur | box data **(null if QrCode is not associated with any box)** | success |

Note how currently the cases 1, 3a, 3b cannot be distinguished.

Also with the new interface design now in case of 3b information about the base/org of the box can be displayed.

### CODE REVIEW

**Comment by pylipp** [comment]
[When scanning the QR code of a box registered at another organisation, instead of showing an error about authz failure, show information about which organisation and base the box actually belongs to (i.e. copy behavior of dropapp).](https://trello.com/c/hSZBiyUd/1466-when-scanning-the-qr-code-of-a-box-registered-at-another-organisation-instead-of-showing-an-error-about-authz-failure-show-infor)

**Comment by pylipp** [comment]
@fhenrich33 I consider this completed from a BE perspective. Please check the first post about the different return types of the `qrCode` query and the `QrCode.box` field. Overall, the logic in `useQrResolver` needs a major rework now :sweat_smile: let me know how it goes for you. I could already adjust `queries.ts` though.
Since the changes in the GraphQL API break FE, I suggest you continue to work on it on the same branch, what do you think?

**Comment by fhenrich33** [comment]
> @fhenrich33 I consider this completed from a BE perspective. Please check the first post about the different return types of the `qrCode` query and the `QrCode.box` field. Overall, the logic in `useQrResolver` needs a major rework now 😅 let me know how it goes for you. I could already adjust `queries.ts` though. Since the changes in the GraphQL API break FE, I suggest you continue to work on it on the same branch, what do you think?

Makes sense, will do!

**Review by HaGuesto — COMMENTED** [review]
Thanks @pylipp! Great PR until now.

# code review
It passes, but some FYI for @fhenrich33 :
- Philipp changed the input variable name for `qrCode` and `qrExists` query. We should check the places where we execute the query.

# functional review
passes as well. This is the query I used to test:

```
{
  qrExists(code: "093f65e080a295f8076b1c5722a46aa2")
  qrCodeFail: qrCode(code: "093f65e080a295f8076b1c5722a46aa2") {
    ... on QrCode {
      id
      box {
        ... on Box {
          labelIdentifier
        }
        ... on InsufficientPermissionError {
          name
        }
        ... on UnauthorizedForBaseError {
          name
          organisationName
        }
      }
    }
    ... on InsufficientPermissionError {
      name
    }
    ... on InsufficientPermissionError {
      name
    }
  }
  qrCodeValid: qrCode (code: "9d7f34fbea04de63b0acdfb874f3ef59") {
    ... on QrCode {
      id
      box {
        ... on Box {
          labelIdentifier
        }
        ... on InsufficientPermissionError {
          name
        }
        ... on UnauthorizedForBaseError {
          name
          organisationName
        }
      }
    }
    ... on InsufficientPermissionError {
      name
    }
    ... on InsufficientPermissionError {
      name
    }
  }
}

```

File: front/src/hooks/useQrResolver.ts [hunk 1]
```diff
@@ -13,18 +13,20 @@ export enum IQrResolverResultKind {
   NOT_ASSIGNED_TO_BOX = "notAssignedToBox",
```
Comment (HaGuesto) [inline front/src/hooks/useQrResolver.ts:13]: I'd suggest to rework these cases:
- NO_BOXTRIBUTE_QR: `data.qrCode.__typename === "ResourceDoesNotExistError"  --> "No box found for this QR code!" toast (Currently line 65)
- NOT_AUTHORIZED_FOR_QR: `data.qrCode.__typename === "InsufficientPermissionError"  --> "You don't have permission to access this box!" toast (Currently line 56)
- NOT_ASSIGNED_TO_BOX: !data?.qrCode?.box --> forwarding to createBox component (Currently line 82)
- NOT_AUTHORIZED_FOR_BOX: data.qrCode.box?.__typename === "InsufficientPermissionError" --> "You don't have permission to access this box!" toast
- NOT_AUTHORIZED_FOR_BASE: data.qrCode.box?.__typename === "UnauthorizedForBaseError" --> show the name of the other base
- SUCCESS --> forwarding to Boxview BoxView
- FAIL --> fallback case for any unexpected error

File: front/src/hooks/useQrResolver.ts [hunk 2]
```diff
@@ -52,7 +57,7 @@ export const useQrResolver = () => {
         })
         .then(({ data, errors }) => {
           if ((errors?.length || 0) > 0) {
-            const errorCode = errors ? errors[0].extensions?.code : undefined;
+            const errorCode = errors ? errors[0]?.extensions?.code : undefined;
             if (errorCode === "FORBIDDEN") {
```
Comment (HaGuesto) [inline front/src/hooks/useQrResolver.ts:61]: the only errorCode left is if an unexpected error happens --> IQrResolverResultKind = FAIL should only be here

File: front/src/hooks/useQrResolver.ts [hunk 3]
```diff
@@ -79,22 +84,44 @@ export const useQrResolver = () => {
...
             return {
               kind: IQrResolverResultKind.NOT_ASSIGNED_TO_BOX,
               qrHash: hash,
             } as IQrResolvedValue;
           }
+          // Handle valid not owned Box cases.
+          // @ts-ignore
+          if (data.qrCode.box?.__typename === "InsufficientPermissionError")
+            return {
+              kind: IQrResolverResultKind.BOX_NO_PERMISSION,
```
Comment (HaGuesto) [inline front/src/hooks/useQrResolver.ts:98]: rename the resultKind to NOT_AUTHORIZED_FOR_BOX

File: front/src/hooks/useQrResolver.ts [hunk 4]
```diff
@@ -79,22 +84,44 @@ export const useQrResolver = () => {
...
+              qrHash: hash,
+              // @ts-ignore
+              box: data?.qrCode?.box,
+            } as IQrResolvedValue;
+
+          // @ts-ignore
+          if (data.qrCode.box?.__typename === "UnauthorizedForBaseError")
+            return {
+              kind: IQrResolverResultKind.BOX_NOT_AUTHORIZED,
```
Comment (HaGuesto) [inline front/src/hooks/useQrResolver.ts:107]: rename the resultKind to NOT_AUTHORIZED_FOR_BASE

File: front/src/hooks/useQrResolver.ts [hunk 5]
```diff
@@ -79,22 +84,44 @@ export const useQrResolver = () => {
               qrHash: hash,
             } as IQrResolvedValue;
           }
+          // @ts-ignore
           if (!data?.qrCode?.box) {
```
Comment (HaGuesto) [inline front/src/hooks/useQrResolver.ts:88]: missing cases are:

- NO_BOXTRIBUTE_QR: `data.qrCode.__typename === "ResourceDoesNotExistError" --> "No box found for this QR code!" toast (Currently line 65)
- NOT_AUTHORIZED_FOR_QR: `data.qrCode.__typename === "InsufficientPermissionError" --> "You don't have permission to access this box!" toast (Currently line 56)

File: front/src/hooks/useQrResolver.ts [hunk 6]
```diff
@@ -79,22 +84,44 @@ export const useQrResolver = () => {
               qrHash: hash,
             } as IQrResolvedValue;
```
Comment (HaGuesto) [inline front/src/hooks/useQrResolver.ts:85]: unexpected graphQL error

File: front/src/hooks/useQrResolver.ts [hunk 7]
```diff
@@ -79,22 +84,44 @@ export const useQrResolver = () => {
...
           return {
             kind: IQrResolverResultKind.SUCCESS,
             qrHash: hash,
+            // @ts-ignore
             box: data?.qrCode?.box,
           } as IQrResolvedValue;
         })
         .catch((err) => {
           triggerError({
             message: "QR code lookup failed. Please wait a bit and try again.",
```
Comment (HaGuesto) [inline front/src/hooks/useQrResolver.ts:122]: unexpected network error

File: front/src/components/QrReader/QrReaderContainer.tsx [hunk 8]
```diff
@@ -34,6 +34,7 @@ function QrReaderContainer({ onSuccess }: IQrReaderContainerProps) {
   const [isMultiBox, setIsMultiBox] = useState(!!qrReaderOverlayState.isMultiBox);
   const [isProcessingQrCode, setIsProcessingQrCode] = useState(false);
   const [isCameraNotPermited, setIsCameraNotPermited] = useState(false);
+  const [boxNotOnwned, setBoxNotOnwned] = useState("");
```
Comment (HaGuesto) [inline front/src/components/QrReader/QrReaderContainer.tsx:37]: pls fix typo

File: front/src/components/QrReader/QrReaderContainer.tsx [hunk 9]
```diff
@@ -73,11 +74,24 @@ function QrReaderContainer({ onSuccess }: IQrReaderContainerProps) {
   const onScan = async (qrReaderResultText: string, multiScan: boolean) => {
     if (!isProcessingQrCode) {
       setIsProcessingQrCode(true);
+      setBoxNotOnwned("");
       const qrResolvedValue: IQrResolvedValue = await resolveQrCode(
         qrReaderResultText,
         multiScan ? "cache-first" : "network-only",
       );
       switch (qrResolvedValue.kind) {
+        case IQrResolverResultKind.BOX_NO_PERMISSION: {
```
Comment (HaGuesto) [inline front/src/components/QrReader/QrReaderContainer.tsx:83]: pls remove this case and rather call a trigger_error in the QRResolver hook that shows than a toast.

File: front/src/hooks/useQrResolver.ts (hunk 3 above)
Comment (HaGuesto) [reply front/src/hooks/useQrResolver.ts:98]: add the triggerError call from line 57

**Review by HaGuesto — CHANGES_REQUESTED** [review]

File: front/src/hooks/useQrResolver.ts [hunk 10]
```diff
@@ -56,68 +56,64 @@ export const useQrResolver = () => {
...
               kind: IQrResolverResultKind.FAIL,
               qrHash: hash,
             } as IQrResolvedValue;
           }
-          // @ts-ignore
-          if (!data?.qrCode?.box) {
+
+          if (data.qrCode.__typename === "ResourceDoesNotExistError") {
             return {
               kind: IQrResolverResultKind.NOT_ASSIGNED_TO_BOX,
```
Comment (pylipp) [inline front/src/hooks/useQrResolver.ts:73]: I think this should be `NO_BOXTRIBUTE_QR` and also show an error toast?

**Comment by fhenrich33** [comment]
@pylipp By the way the tests that are failing doesn't reflect what actually happens in the real app, I didn't figure out a way to fix them yet. cc @HaGuesto 

The failing tests:

```
         3.4.3.5 QrReaderMultiBox.test.tsx
            AssertionError: expected "spy" to be called with arguments: [ ObjectContaining{…} ]

            Received:

            1st spy call:

            Array [
            -   ObjectContaining {
            -     "message": StringMatching /have permission to access this box/i,
            +   Object {
            +     "message": "QR code lookup failed. Please wait a bit and try again.",
                },
            ]
        3.4.8.7 ResolveHash.test.tsx
            TestingLibraryElementError: Unable to find an element by: [data-testid="ReturnScannedQr"]
        3.4.2.2 QrReaderOverlay.test.tsx
            TestingLibraryElementError: Unable to find role="heading" and name "/bases/1/boxes/123"
        3.4.2.5b QrReaderOverlay.test.tsx 
            AssertionError: expected "spy" to be called with arguments: [ ObjectContaining{…} ]

            Received:

            1st spy call:

            Array [
            -   ObjectContaining {
            -     "message": StringMatching /No box found for this QR code/i,
            +   Object {
            +     "message": "QR code lookup failed. Please wait a bit and try again.",
                },
            ]
```

File: front/src/hooks/useQrResolver.ts (hunk 10 above)
Comment (fhenrich33) [reply front/src/hooks/useQrResolver.ts:73]: This is handled at `/front/src/components/QrReader/QrReaderContainer.tsx`

```ts
case IQrResolverResultKind.NOT_ASSIGNED_TO_BOX: {
  if (!multiScan) {
    onSuccess();
    navigate(`/bases/${baseId}/boxes/create/${qrResolvedValue?.qrHash}`);
  } else {
    triggerError({
      message: "No box associated to this QR code!",
    });
    setIsProcessingQrCodeDelayed(false);
  }
  break;
}
```

Comment (pylipp) [reply front/src/hooks/useQrResolver.ts:73]: yes it works :+1: But I end up on an empty page `http://localhost:3000/bases/1/boxes/create/13f12820c8010f2f7349962930e6bf4` (I used the last one of the invalid QR codes from the repo readme).
I'd expect to stay on `/qrread`

Comment (fhenrich33) [reply front/src/hooks/useQrResolver.ts:73]: Huh, this should've been handled at `front/src/hooks/useQrResolver.ts`

```ts
const resolveQrCode = useCallback(
  async (qrCodeUrl: string, fetchPolicy: FetchPolicy): Promise<IQrResolvedValue> => {
    setLoading(true);
    const extractedQrHashFromUrl = extractQrCodeFromUrl(qrCodeUrl);

    if (!extractedQrHashFromUrl) {
      triggerError({
        message: "This is not a Boxtribute QR code!",
      });

      setLoading(false);
      return { kind: IQrResolverResultKind.NOT_BOXTRIBUTE_QR } as IQrResolvedValue;
    }

    const qrResolvedValue: IQrResolvedValue = await resolveQrHash(
      extractedQrHashFromUrl,
      fetchPolicy,
    );

    setLoading(false);
    return qrResolvedValue;
  },
  [resolveQrHash, triggerError],
);
```

Cheking this out.

**Comment by HaGuesto** [comment]
@pylipp this PR is ready to be merge. Pls do so when you think it is the right time.

**Review by HaGuesto — APPROVED** [review]

**Pull request merged**